src/Microsoft.Performance.SDK/Processing/ITableDiscoverer.cs

```diff
@@ -0,0 +1,309 @@
+
+            return CreateForNamespace(@namespace, new[] { assembly, });
+        }
+
+        /// <summary>
+        ///     Creates a new <see cref="ITableProvider"/> instance that will
+        ///     discover tables in the given namespace in the given assembly.
+        /// </summary>
+        /// <param name="namespace">
+        ///     The namespace to search for tables.
+        /// </param>
+        /// <param name="assemblies">
+        ///     The <see cref="Assembly"/> instances to search for tables. Any null elements in
+        ///     this parameter are ignored. Any duplicate assemblies are ignored.
+        /// </param>
+        /// <returns>
+        ///     A new <see cref="ITableProvider"/> that discovers tables in the
+        ///     specified <paramref name="namespace"/> in the specified <paramref name="assembly"/>.
+        /// </returns>
+        /// <exception cref="System.ArgumentNullException">
+        ///     <paramref name="namespace"/> is <c>null</c>.
+        ///     - or -
+        ///     <paramref name="assemblies"/> is <c>null</c>.
+        /// </exception>
+        /// <exception cref="System.ArgumentException">
+        ///     <paramref name="assemblies"/> has no non-null elements.
+        /// </exception>
+        public static ITableProvider CreateForNamespace(string @namespace, Assembly[] assemblies)
+        {
+            Guard.NotNull(@namespace, nameof(@namespace));
+            Guard.NotNull(assemblies, nameof(assemblies));
+
+            var set = new HashSet<Assembly>();
+
+            for (var i = 0; i < assemblies.Length; ++i)
+            {
+                if (assemblies[i] != null)
+                {
+                    set.Add(assemblies[i]);
+                }
+            }
+
+            if (set.Count == 0)
+            {
+                throw new ArgumentException(
+                    $"{nameof(assemblies)} must contain at least one non-null element.",
+                    nameof(assemblies));
+            }
+
+            return new NamespaceTableDiscoverer(@namespace, set);
+        }
+
+        private static ISet<DiscoveredTable> DoAssemblyDiscovery(
+            IEnumerable<Assembly> assemblies,
+            Func<Type, bool> typeFilter,
+            ISerializer tableConfigSerializer)
+        {
+            Debug.Assert(assemblies != null);
+            Debug.Assert(typeFilter != null);
+            Debug.Assert(assemblies.Any());
+
+            var s = new HashSet<DiscoveredTable>();
+            foreach (var t in assemblies.SelectMany(x => x.GetTypes()).Where(typeFilter))
+            {
+                if (TableDescriptorFactory.TryCreate(t, tableConfigSerializer, out var isInternal, out var td, out var buildTable))
+                {
+                    s.Add(new DiscoveredTable(td, buildTable, isInternal));
+                }
+            }
+
+            return s;
+        }
+
+        private sealed class AssemblyTableDiscoverer
```

Comment:
why are these classes private? What's the benefit of having the above static methods instead of these classes just having public constructors?

Reply:
This allows the end user to depend only on the interfaces. I wanted to avoid having concrete types for this in case we needed to refactor implementations, or add additional functionality.